File: tests/python/test_fir_filter.py
import numpy as np
import pytest
from fir_filter import HalfBand, InverseSinc

@pytest.fixture
def inversesinc_filter():
    return InverseSinc()

@pytest.fixture
def halfband_filter():
    return HalfBand()

def test_inversesinc_valid_output(inversesinc_filter):
    b = inversesinc_filter(0.4, 0.025)
    assert(np.all(b[:len(b)//2] == b[-1:-len(b)//2:-1])), "Filter must be symetric."

def test_inversesinc_order_increase(inversesinc_filter):
    b_lower = inversesinc_filter(0.4, 0.025)
    b_higher = inversesinc_filter(0.45, 0.025)
    assert(len(b_lower) < len(b_higher)), "Higher bandpass equals higher filter order."

    b_lower = inversesinc_filter(0.4, 0.25)
    b_higher = inversesinc_filter(0.4, 0.025)
    assert(len(b_lower) < len(b_higher)), "Lower error equals higher filter order."

def test_inversesinc_too_high_order(inversesinc_filter):
    with pytest.raises(ValueError, match="InverseSinc.__call__: Too high order of the filter."):
        inversesinc_filter(0.49, 0.0025)

def test_inversesinc_invalid_input(inversesinc_filter):
    f_passes = [-0.1, 0.6]
    for f_pass in f_passes:
        with pytest.raises(ValueError, match="InverseSinc.__call__: Passband must lie between 0.0 and 0.5."):
            inversesinc_filter(f_pass, 0.025)

def test_halfband_valid_output(halfband_filter):
    coeffs = halfband_filter(60.0, 0.2)
    assert(np.isclose(coeffs[len(coeffs)//2], 0.5, atol=1e-5)), "Center coefficient should be close to 0.5."
    assert(np.all(coeffs[1::2][coeffs[1::2] != coeffs[len(coeffs)//2]]) == 0), "Every odd coefficient should be 0, except center coefficient."
    assert(np.all(coeffs[:len(coeffs)//2] == coeffs[-1:-len(coeffs)//2:-1])), "Filter must be symetric."

def test_halfband_order_increase(halfband_filter):
    coeffs_lower = halfband_filter(60.0, 0.1)
    coeffs_higher = halfband_filter(60.0, 0.2)
    assert(len(coeffs_lower) < len(coeffs_higher)), "Higher bandpass equals higher filter order."

    coeffs_lower = halfband_filter(60.0, 0.2)
    coeffs_higher = halfband_filter(80.0, 0.2)
    assert(len(coeffs_lower) < len(coeffs_higher)), "Higher attenuation equals higher filter order."

def test_halfband_too_high_order(halfband_filter):
    with pytest.raises(ValueError, match="HalfBand.__call__: Too high order of the filter."):
        halfband_filter(100.0, 0.23)

def test_halfband_invalid_input(halfband_filter):
    f_passes = [-0.1, 0.3]
    for f_pass in f_passes:
        with pytest.raises(ValueError, match="Passband must lie between 0.0 and 0.25."):
            halfband_filter(60.0, f_pass)

File: tests/cpp/test_firFilter.cpp
#include <gtest/gtest.h>
#include "firFilter.h"

using namespace std;

class TestFIR : public ::testing::Test {
    protected:
        HalfBand halfband;
        InverseSinc inverseSinc;
};

TEST_F(TestFIR, inverseSincValidOutput) {
    // Compute the result.
    vector<double> b = inverseSinc(0.4, 0.025);

    // Test the result.
    for (uint i = 0; i < b.size() / 2; i++) { ASSERT_EQ(b[i], b[b.size()-1-i]) << "Filter must be symetric."; }
}

TEST_F(TestFIR, inverseSincOrderIncrease) {
    // Compute the result.
    vector<double> bLower = inverseSinc(0.4, 0.025);
    vector<double> bHigher = inverseSinc(0.45, 0.025);

    // Test the result.
    ASSERT_LT(bLower.size(), bHigher.size()) << "Higher bandpass equals higher filter order.";

    // Compute the result.
    bLower = inverseSinc(0.4, 0.25);
    bHigher = inverseSinc(0.4, 0.025);

    // Test the result.
    ASSERT_LT(bLower.size(), bHigher.size()) << "Lower error equals higher filter order.";
}

TEST_F(TestFIR, inverseSincTooHighOrder) {
    // Too high order of the filter.
    EXPECT_THROW(inverseSinc(0.49, 0.0025), length_error);
}

TEST_F(TestFIR, inverseSincInvalidInput) {
    double Fpass = -0.1; // Passband must lie between 0.0 and 0.5.
    EXPECT_THROW(halfband(60.0, Fpass), invalid_argument);

    Fpass = 0.6; // Passband must lie between 0.0 and 0.5.
    EXPECT_THROW(halfband(60.0, Fpass), invalid_argument);
}

TEST_F(TestFIR, halfBandValidOutput) {
    // Compute the result.
    vector<double> coeffs = halfband(120.0, 0.2);

    // Test the result.
    ASSERT_NEAR(coeffs[coeffs.size()/2], 0.5, 5e-3) << "Center coefficient should be close to 0.5.";
    for (uint i = 0; i < coeffs.size(); i++) {
        if (i % 2 && i != coeffs.size() / 2) {
            ASSERT_EQ(coeffs[i], 0.0) << "Every odd coefficient should be 0, except center coefficient.";
        }
        if (i < coeffs.size() / 2) {
            ASSERT_EQ(coeffs[i], coeffs[coeffs.size()-1-i]) << "Filter must be symetric.";
        }
    }
}

TEST_F(TestFIR, halfBandOrderIncrease) {
    // Compute the result.
    vector<double> coeffsLower = halfband(120.0, 0.1);
    vector<double> coeffsHigher = halfband(120.0, 0.2);

    // Test the result.
    ASSERT_LT(coeffsLower.size(), coeffsHigher.size()) << "Higher bandpass equals higher filter order.";

    // Compute the result.
    coeffsLower = halfband(120.0, 0.2);
    coeffsHigher = halfband(130.0, 0.2);

    // Test the result.
    ASSERT_LT(coeffsLower.size(), coeffsHigher.size()) << "Higher attenuation equals higher filter order.";
}

TEST_F(TestFIR, halfBandTooHighOrder) {
    // Too high order of the filter.
    EXPECT_THROW(halfband(270.0, 0.24), length_error);
}

TEST_F(TestFIR, halfBandInvalidInput) {
    double Fpass = -0.1; // Passband must lie between 0.0 and 0.25.
    EXPECT_THROW(halfband(120.0, Fpass), invalid_argument);

    Fpass = 0.3; // Passband must lie between 0.0 and 0.25.
    EXPECT_THROW(halfband(120.0, Fpass), invalid_argument);
}
